// @file
// @brief Server RPC proto file for IPv6.
//
// ----------------------------------------------------------------
//  Copyright (c) 2016 by cisco Systems, Inc.
//  All rights reserved.
// -----------------------------------------------------------------
//
//

syntax = "proto3";
package service_layer;

import "sl_common_types.proto";
import "sl_route_common.proto";

// @defgroup SLRouteIPv6Oper
// @ingroup Route
// Defines RPC calls for IPv6 route changes and VRF registration.
// This service declares both the Vrf Registration, as well as adding, deleting
// and getting IPv6 routes.
// @{
service SLRoutev6Oper {
    //@addtogroup SLRouteIPv6Oper
    //@{
    //;

    //
    // Global Route operations
    //

    // Used to retrieve Global Route information
    rpc SLRoutev6GlobalsGet(SLRouteGlobalsGetMsg) returns (SLRouteGlobalsGetMsgRsp);

    // Used to retrieve Global Route Stats
    rpc SLRoutev6GlobalStatsGet(SLRouteGlobalStatsGetMsg) returns (SLRouteGlobalStatsGetMsgRsp);

    //
    // VRF registration operations
    //

    // SLVrfRegMsg.Oper = SL_REGOP_REGISTER:
    //     VRF registration: Sends a list of VRF registration messages
    //     and expects a list of registration responses.
    //     A client Must Register a VRF BEFORE routes can be added/modified in
    //    the associated VRF.
    //
    // SLVrfRegMsg.Oper = SL_REGOP_UNREGISTER:
    //     VRF Un-registeration: Sends a list of VRF un-registration messages
    //     and expects a list of un-registration responses.
    //     This can be used to convey that the client is no longer interested
    //     in this VRF. All previously installed routes would be lost.
    //
    // SLVrfRegMsg.Oper = SL_REGOP_EOF:
    //     VRF End Of File message.
    //     After Registration, the client is expected to send an EOF
    //     message to convey the end of replay of the client's known objects.
    //     This is especially useful under certain restart scenarios when the
    //     client and the server are trying to synchronize their Routes.
    rpc SLRoutev6VrfRegOp(SLVrfRegMsg) returns (SLVrfRegMsgRsp);

    // VRF get. Used to retrieve VRF attributes from the server.
    rpc SLRoutev6VrfRegGet(SLVrfRegGetMsg) returns (SLVrfRegGetMsgRsp);

    // Used to retrieve VRF Stats from the server.
    rpc SLRoutev6VrfGetStats(SLVrfRegGetMsg) returns (SLVRFGetStatsMsgRsp);

    //
    // Route operations
    //

    // SLRoutev6Msg.Oper = SL_OBJOP_ADD:
    //     Route add. Fails if the route already exists.
    //
    // SLRoutev6Msg.Oper = SL_OBJOP_UPDATE:
    //     Route update. Creates or updates the route.
    //
    // SLRoutev6Msg.Oper = SL_OBJOP_DELETE:
    //     Route delete. The route path is not necessary to delete the route.
    rpc SLRoutev6Op(SLRoutev6Msg) returns (SLRoutev6MsgRsp);

    // Retrieves route attributes.
    rpc SLRoutev6Get(SLRoutev6GetMsg) returns (SLRoutev6GetMsgRsp);

    //
    // Stream Route operations
    //

    // SLRoutev6Msg.Oper = SL_OBJOP_ADD:
    //     Route add. Fails if the route already exists.
    //
    // SLRoutev6Msg.Oper = SL_OBJOP_UPDATE:
    //     Route update. Creates or updates the route.
    //
    // SLRoutev6Msg.Oper = SL_OBJOP_DELETE:
    //     Route delete. The route path is not necessary to delete the route.
    rpc SLRoutev6OpStream(stream SLRoutev6Msg) returns (stream SLRoutev6MsgRsp);

    // Retrieves route attributes.
    rpc SLRoutev6GetStream(stream SLRoutev6GetMsg) returns (stream SLRoutev6GetMsgRsp);
    //@}
}

// IPv6 route
message SLRoutev6 {
    // IPv6 Prefix
    //0:: - ffff:ffff:ffff:ffff:ffff:ffff:ffff:ffff
    bytes Prefix = 1;

    // IPv6 prefix length [0-128]
    uint32 PrefixLen = 2;

    // Common route attributes
    SLRouteCommon RouteCommon = 3;

    // List of route paths for a particular route.
    // Specifying more than one path is allowed for ECMP/UCMP cases
    repeated SLRoutePath PathList = 4;
}

// List of routes for bulk download
message SLRoutev6Msg {
    // Route Object Operations
    SLObjectOp Oper = 1;

    // Correlator. This can be used to correlate replies with requests.
    // The Server simply reflects this field back in the reply.
    uint64 Correlator = 2;

    // VRF name.
    string VrfName = 3;

    // List of routes for the VRF specified above
    repeated SLRoutev6 Routes = 4;
}

// IPv6 route result, uniquely identified by the Prefix/PrefixLen pair
message SLRoutev6Res {
    // Corresponding error code
    SLErrorStatus ErrStatus = 1;

    // IPv6 Prefix
    bytes Prefix = 2;

    // IPv6 prefix length, [0-32]
    uint32 PrefixLen = 3;
}

// IPv6 bulk route result status
message SLRoutev6MsgRsp {
    // Correlator. This can be used to correlate replies with requests.
    // The Server simply reflects this field back in the reply.
    uint64 Correlator = 1;

    // VRF name (matches the VRF name of the original operation)
    string VrfName = 2;

    // Summary result of the bulk operation (refer to enum SLErrorStatus)
    // In general, the StatusSummary implies one of 3 things:
    // 1. SL_SUCCESS: signifies that the entire bulk operation was successful.
    //         In this case, the Results list is empty.
    // 2. SL_SOME_ERR: signifies that the operation failed for one or more
    //         entries. In this case, Results holds the result for
    //         each individual entry in the bulk.
    // 3. SL_RPC_XXX: signifies that the entire bulk operation failed.
    //         In this case, the Results list is empty.
    SLErrorStatus StatusSummary = 3;

    // In case of errors, this field indicates which entry in the bulk was
    // erroneous.
    repeated SLRoutev6Res Results = 4;
}

// Used to retrieve route attributes
message SLRoutev6GetMsg {
    // Correlator. This can be used to correlate replies with requests.
    // The Server simply reflects this field back in the reply.
    uint64 Correlator = 1;

    // VRF name.
    // If the Key is not specified, then request up to the first 
    // 'EntriesCount' entries.
    string VrfName = 2;

    // IPv6 Prefix
    bytes Prefix = 3;

    // IPv6 prefix length, [0-32]
    uint32 PrefixLen = 4;

    // Number of entries requested
    uint32 EntriesCount = 5;

    // if GetNext is FALSE:
    //     request up to 'EntriesCount' entries starting from the key
    // If GetNext is TRUE, or if the key exact match is not found:
    //     request up to 'EntriesCount' entries starting from the key's next
    bool GetNext = 6;
}

// Gt Route message response
message SLRoutev6GetMsgRsp {
    // Correlator. This can be used to correlate replies with requests.
    // The Server simply reflects this field back in the reply.
    uint64 Correlator = 1;

    // End Of File.
    // When set to True, it indicates that the server has returned M, where 
    // M < N, of the original N requested Entries.
    bool Eof = 2;

    // VRF name.
    string VrfName = 3;

    // Status of the Get operation
    SLErrorStatus ErrStatus = 4;

    // Returned entries as requested in the Get operation.
    // if ErrStatus is SL_SUCCESS, Entries contains the info requested
    repeated SLRoutev6 Entries = 5;
}

//@}
